core/src/main/java/org/apache/carbondata/core/scan/executor/util/RestructureUtil.java

```diff
@@ -119,14 +119,18 @@
         for (CarbonDimension tableDimension : tableComplexDimension) {
           if (isColumnMatches(isTransactionalTable, queryDimension.getDimension(),
               tableDimension)) {
-            ProjectionDimension currentBlockDimension = null;
+            ProjectionDimension currentBlockDimension;
             // If projection dimension is child of struct field and contains Parent Ordinal
             if (null != queryDimension.getDimension().getComplexParentDimension()) {
               currentBlockDimension = new ProjectionDimension(queryDimension.getDimension());
             } else {
               currentBlockDimension = new ProjectionDimension(tableDimension);
+              // for complex dimension update datatype, set scale and precision by traversing.
+              // Spark requires the resultant row with latest datatype,
+              // update the dimension here to collect the resultant rows with updated datatype.
+              fillNewDatatypesForComplexChildren(currentBlockDimension.getDimension(),
+                  queryDimension.getDimension());
             }
-            // TODO: for complex dimension set scale and precision by traversing
             // the child dimensions
```

Comment:
remove this here and add along with new comment

Reply:
Done